include/nbl/builtin/hlsl/fft/common.hlsl

```diff
@@ -1,60 +1,128 @@
+    return numberOfComplexElements * (halfFloats ? sizeof(complex_t<float16_t>) : sizeof(complex_t<float32_t>));
+}
+
 // Computes the kth element in the group of N roots of unity
 // Notice 0 <= k < N/2, rotating counterclockwise in the forward (DIF) transform and clockwise in the inverse (DIT)
 template<bool inverse, typename Scalar>
 complex_t<Scalar> twiddle(uint32_t k, uint32_t halfN)
 {
     complex_t<Scalar> retVal;
-    const Scalar kthRootAngleRadians = numbers::pi<Scalar> * Scalar(k) / Scalar(halfN);
-    retVal.real( cos(kthRootAngleRadians) );
-    if (! inverse)
-        retVal.imag( sin(-kthRootAngleRadians) );
+    const Scalar kthRootAngleRadians = numbers::pi<Scalar> *Scalar(k) / Scalar(halfN);
+    retVal.real(cos(kthRootAngleRadians));
+    if (!inverse)
+        retVal.imag(sin(-kthRootAngleRadians));
     else
-        retVal.imag( sin(kthRootAngleRadians) );
-    return retVal;                         
+        retVal.imag(sin(kthRootAngleRadians));
+    return retVal;
 }
 
-template<bool inverse, typename Scalar> 
-struct DIX 
-{ 
+template<bool inverse, typename Scalar>
+struct DIX
+{
     static void radix2(NBL_CONST_REF_ARG(complex_t<Scalar>) twiddle, NBL_REF_ARG(complex_t<Scalar>) lo, NBL_REF_ARG(complex_t<Scalar>) hi)
     {
         plus_assign< complex_t<Scalar> > plusAss;
         //Decimation in time - inverse           
         if (inverse) {
             complex_t<Scalar> wHi = twiddle * hi;
             hi = lo - wHi;
-            plusAss(lo, wHi);            
+            plusAss(lo, wHi);
         }
         //Decimation in frequency - forward   
         else {
             complex_t<Scalar> diff = lo - hi;
             plusAss(lo, hi);
-            hi = twiddle * diff; 
+            hi = twiddle * diff;
         }
-    }                                              
+    }
 };
 
 template<typename Scalar>
 using DIT = DIX<true, Scalar>;
 
 template<typename Scalar>
 using DIF = DIX<false, Scalar>;
+
+// ------------------------------------------------- Utils ---------------------------------------------------------
+// 
+// Util to unpack two values from the packed FFT X + iY - get outputs in the same input arguments, storing x to lo and y to hi
+template<typename Scalar>
+void unpack(NBL_REF_ARG(complex_t<Scalar>) lo, NBL_REF_ARG(complex_t<Scalar>) hi)
+{
+    complex_t<Scalar> x = (lo + conj(hi)) * Scalar(0.5);
+    hi = rotateRight<Scalar>(lo - conj(hi)) * Scalar(0.5);
+    lo = x;
+}
+
+
 }
 }
 }
 
+#ifdef __HLSL_VERSION
+
+namespace nbl 
+{
+namespace hlsl
+{
+namespace fft 
+{
+
+// ------------------------------------------------- Utils ---------------------------------------------------------
+
+// Bit-reverses T as a binary string of length given by Bits
+template<typename T, uint16_t Bits NBL_FUNC_REQUIRES(is_integral_v<T> && Bits <= sizeof(T) * 8)
+T bitReverse(T value)
+{
+    return spirv::bitReverse<uint32_t>(value) >> (sizeof(T) * 8 - Bits);
```

Comment:
`return spirv::bitReverse<T>(value) >> promote<T>(sizeof(T) * 8 - Bits);`

Reply:
also I don't get why you made it C++ only

Reply:
works for unsigned vectors only, should I require T be unsigned and be done with that?

Reply:
> works for unsigned vectors only, should I require T be unsigned and be done with that?

yes require unsigned, it makes sense that way